### DISCUSSION TITLE
Changing border rendering of terminals

### DISCUSSION THREAD

**Discussion by SquidDev** [cc-tweaked/CC-Tweaked#1911] [General]
This isn't an issue or PR, as I'm not planning to do it anything further about this, but felt it would be useful to demo some work/fiddling I've done:

---

CC draws the border margin using the adjacent background colour. This works most of the time, but looks very odd when drawing inverted teletext characters. For instance, here is `\x99` rendered with white-on-black and black-on-white:

![A screenshot of two teletext characters. Each character is three dots on a 2x3 pixel grid, forming an arrow pointing to the right. The first is drawn with white text on a black background and looks normal, the second is drawn with black text on a white background, but with another white block on the left (due to the margin), causing it to look very weird.](https://github.com/user-attachments/assets/a0a2e90c-72cc-4fb8-a476-5f08ca1c4037)

There was some discussion a long time ago [on the original ComputerCraft repo](https://github.com/dan200/ComputerCraft/issues/51)  about this, and it has up a couple of times since on CC:T's own issue tracker (e.g. #722).

There, a couple of options were proposed,  which I've had a fiddle with implementing. All three screenshots are the result of running the following program (yet, this is very lazy and non-representative):

```lua
for i =  1, 19 do
  term.setCursorPos(1, i)
  if i % 2 == 0 then
    term.setBackgroundColor(colours.white)
    term.setTextColor(colours.black)
    term.write("\x8e")
  else
    term.setBackgroundColor(colours.black)
    term.setTextColor(colours.white)
    term.write("\x99")
  end
end
```

## Keep as is
![Screenshot from 2024-07-25 21-05-55](https://github.com/user-attachments/assets/d57d2301-dc85-4c57-9793-eca7bb7c7939)
![Screenshot from 2024-07-25 21-37-06](https://github.com/user-attachments/assets/1b3e28b9-1968-495f-83be-0bf5500f4018)

## Extend the teletext characters to the edge
![Screenshot from 2024-07-25 21-05-33](https://github.com/user-attachments/assets/fe535df5-1db7-4b0e-b141-92284a85eed9)

![Screenshot from 2024-07-25 21-36-27](https://github.com/user-attachments/assets/2abefbef-c273-4d8b-8647-415cb7503075)

## Black border
![Screenshot from 2024-07-25 21-10-17](https://github.com/user-attachments/assets/dea0a99e-3872-44dc-93a4-a0b973d5a4a4)

## Conclusion
I was originally quite hopeful about "Extend the teletext characters to the edge". This is what I'd suggested [in the original issue](https://github.com/dan200/ComputerCraft/issues/51#issuecomment-311936484), but looking at it in practice, I'm less sure about it. It works pretty well for art, and I suspect will be fine for GUIs (as they're fairly blocky), but still feels a little strange for the more complex characters.

I do still prefer the black border in most cases. However, I do agree with Dan's original complaints in https://github.com/dan200/ComputerCraft/pull/242 that it looks a little odd for programs like `paint`.

**Comment by Lupus590** [comment]
I like the black border. It'll be mostly the same feel and won't make the border pixels seem like they are extra big in odd dimensions.
